cdc/generation.cc

```diff
@@ -123,176 +123,198 @@ const std::vector<token_range_description>& topology_description::entries() cons
+     * when going from 100 to 10 on the ring.
+     */
+    dht::token find_first_token_for_shard(
+            const dht::sharder& sharder, size_t shard_idx) const {
+        auto t = next_token(start);
+        if (sharder.shard_of(t) != shard_idx) {
+            t = sharder.token_for_next_shard(start, shard_idx);
         }
-
-        spots_to_fill += entry.streams.size();
-    }
-
-    auto schema = schema_builder("fake_ks", "fake_table")
-        .with_column("stream_id", bytes_type, column_kind::partition_key)
-        .build();
-
-    auto quota = std::chrono::seconds(spots_to_fill / 2000 + 1);
-    auto start_time = std::chrono::system_clock::now();
-
-    // For each pair (i, j), 0 <= i < streams.size(), 0 <= j < streams[i].size(),
-    // try to find a stream (stream[i][j]) such that the token of this stream will get mapped to this stream
-    // (refer to the comments above topology_description's definition to understand how it describes the mapping).
-    // We find the streams by randomly generating them and checking into which pairs they get mapped.
-    // NOTE: this algorithm is temporary and will be replaced after per-table-partitioner feature gets merged in.
-    repeat([&] {
-        for (int i = 0; i < 500; ++i) {
-            auto stream_id = make_random_stream_id();
-            auto token = dht::get_token(*schema, stream_id.to_partition_key(*schema));
-
-            // Find the token range into which our stream_id's token landed.
-            auto it = std::lower_bound(tokens.begin(), tokens.end(), token);
-            auto& entry = entries[it != tokens.end() ? std::distance(tokens.begin(), it) : 0];
-
-            auto shard_id = dht::shard_of(entry.streams.size(), entry.sharding_ignore_msb, token);
-            assert(shard_id < entry.streams.size());
-
-            if (!entry.streams[shard_id].is_set()) {
-                --spots_to_fill;
-                entry.streams[shard_id] = stream_id;
-            }
+        if (!t.is_maximum()) {
+            // This means we have found a token for shard shard_idx before 2^63
+            return t;
         }
-
-        if (!spots_to_fill) {
-            return stop_iteration::yes;
+        // No token owned by shard shard_idx was found in (this.start, 2^63 - 1]
+        // so we have to search in [-2^63, this.end]
+        t = smallest_token();
+        if (sharder.shard_of(t) != shard_idx) {
+            t = sharder.token_for_next_shard(t, shard_idx);
         }
+        return std::min(t, end);
+    }
+};
 
-        auto now = std::chrono::system_clock::now();
-        auto passed = std::chrono::duration_cast<std::chrono::seconds>(now - start_time);
-        if (passed > quota) {
-            return stop_iteration::yes;
+/*
+ * Represents a non wrap around vnode.
+ * This means a token range (this.start, this.end].
+ * For example (10, 100].
+ */
+struct non_wrap_around_vnode final {
+    dht::token start;
```

Comment:
I have to say I still think this can be simplified by using the partition_range_vector_sharder. You're recreating that machinery, with bugs.

You can write something like this.

```c++
auto wrapping_vnode = wrapping_range(start_token, end_token);
auto unwrapped_vnode = wrapping_vnode.unwrap(); // this is a vector
auto token_sharder = ring_position_range_vector_sharder(.., unwrapped_vnode);
while (auto next = token_sharder.next(); nr_tokens < nr_cpus && next) {
    tokens[next->shard] (next->rangebegin());
    ++nr_tokensl
}
```

Unfortunately, what we have now works with ring_positions, not tokens. So you'll need a schema. Is one handy here?

Reply:
Nope. No schema because the generation is shared across all tables.

Reply:
The code you proposed is incorrect the same way current implementation is. One of the returned ranges will start with dht::minnimum_token which is not valid.

If you insist then I will just use the sharder. I don't care about the matter that much.

Reply:
BTW you're way to optimistic about how simple the code will be:
1. wrapping_range.unwrap() returns pair of wrapping_ranges (not nonwrapping_ranges) and if you call it on a range that's actually not wrapping it will return (-inf, start] + [end, inf) so we can't call the same code for wrapping and nonwrapping vnodes.

Reply:
@haaawk, I don't insist on using the range vector sharder, but I insist on using simple code. The code you presented isn't simple. Either simplify it, or use more infrastructure, or add more infrastructure and use it. This is not the place to play token games.

I agree the partition_range_vector_sharder isn't optimal due to its use of infinities and since it requires a schema. Perhaps a new sharder can be used. This separates the concerns.

Reply:
I dealt with the problem of schema by turning ring_position_range_sharder into token_range_sharder but the code does not look simple at all:

```
        dht::sharder sharder(shard_count, ignore_msb);
        dht::token_range_vector_sharder token_sharder(sharder, std::move(vnode));
        while (auto next = token_sharder.next()) {
            auto stream_token = end;
            auto& start = next->ring_range.start();
            if (!start || start->value().is_minimum()) {
                auto next_token = dht::from_int64(std::numeric_limits<int64_t>::min() + 1);
                if (next->ring_range.contains(next_token, dht::tri_compare)) {
                    desc.streams[next->shard] = next_token;
                }
            } else {
                if (start->is_inclusive()) {
                    desc.streams[next->shard] = start->value();
                } else {
                    int64_t token_value = dht::to_int64(start->value());
                    if (token_value < std::numeric_limits<int64_t>::max())) {
                        auto next_token = dht::from_int64(token_value + 1);
                        if (next->ring_range.contains(next_token, dht::tri_compare)) {
                            desc.streams[next->shard] = next_token;
                        }
                    }
                }
            }
        }
```

Reply:
I think I will just try to simplify my previous attempt.